content/blog/2020/08/2020-08-29-git-performance-improvement-phase3.adoc

```diff
@@ -0,0 +1,126 @@
+
+* Size of the aggregated objects is the *dominant* player in determining the execution time for a git fetch
+* Number of branches and Number of tags play a similar role but are strongly overshadowed by size of repository
+* Number of commits has a negligible effect on the performance of running git fetch
+
+*After running these experiments from Stage-1 to Stage-3, we developed a solution called the `GitToolChooser` which is explained in the next stage*
+
+=== Stage 4: Enabling faster checkouts with an automatic git tool chooser
+
+This feature takes the responsibility of choosing the optimal implementation from the *user* and hands it to the *plugin*. It takes the decision of recommending an implementation on the basis of the size of the repository. link:/blog/2020/07/2020-07-29-git-performance-improvement-phase2.adoc[Here is how it works].
+
+image:/images/post-images/2020/07-git-plugin-performance-improvement/git-perf-improv.png[title="Faster checkouts with GitToolChooser" role="center", width=938]
+
+The image above depicts the performance enhancements we have performed over the course of the GSoC project. These improvements have enabled the checkout step to be finished *within half of what it used to take* earlier in some cases.
+
+Let's talk about performance improvements in two parts.
+
+== Major Performance Improvements
+
+image:/images/post-images/2020/07-git-plugin-performance-improvement/git-major-perf.png[Major performance enhancements, align='center', width=400]
+
+Building Tensorflow (~800 MiB) using a Jenkins pipeline, there is over 50% reduction in overall time spent in completing a job! 
+The result is consistent multiple platforms.
+
+The reason for such a decrease is the fact that `JGit` degrades in performance when we are talking about large sized repositories. Since the GitToolChooser is aware of this fact, it chooses to recommend `command line git` instead which saves the user some time.
+
+== Minor Performance Improvements
+
+*Note: Enable JGit before using the new performance features to let GitToolChooser work with more options* -> link:https://plugins.jenkins.io/git-client/#enabling-jgit[Here's how]
+
+image:/images/post-images/2020/07-git-plugin-performance-improvement/git-minor-perf.png[title="Minor performance enhancements" role="center", width=400]
+
+Building the git plugin (~ 20 MiB) using a Jenkins pipeline, there is a drop of a *second* across all platforms when performance enhancement is enabled. Also, eliminating a redundant fetch reduces unnecessary load on git servers.
+
+The reason for this change is the fact that `JGit` performs better than `command line git` for small sized repositories (<50MiB) as an already warmed up JVM favors the native Java implementation.
+
+== Release
+
+* Git Plugin 4.4.0 -> https://github.com/jenkinsci/git-plugin/releases/tag/git-4.4.0
```

Comment:
Hyperlink the text rather than showing the full link to the user
```suggestion
* link:https://github.com/jenkinsci/git-plugin/releases/tag/git-4.4.0[Git Plugin 4.4.0]
```